In [1]:
import pandas as pd
import matplotlib as plt

In [2]:
happiness = pd.read_csv("DataSets/2018-2019+Happiness.csv")
airbnb = pd.read_csv("DataSets/airbnb.csv")
ethereum = pd.read_csv("DataSets/Ethereum+Historical+Data.csv")
salaries = pd.read_csv("DataSets/salaries.csv")

In [3]:
salaries.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8805 entries, 0 to 8804
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   work_year           8805 non-null   int64 
 1   experience_level    8805 non-null   object
 2   employment_type     8805 non-null   object
 3   job_title           8805 non-null   object
 4   salary              8805 non-null   int64 
 5   salary_currency     8805 non-null   object
 6   salary_in_usd       8805 non-null   int64 
 7   employee_residence  8805 non-null   object
 8   remote_ratio        8805 non-null   int64 
 9   company_location    8805 non-null   object
 10  company_size        8805 non-null   object
dtypes: int64(4), object(7)
memory usage: 756.8+ KB


In [4]:
salaries.work_year # Carga todos los datos de una columna, en este caso work_year

0       2023
1       2023
2       2023
3       2023
4       2023
        ... 
8800    2020
8801    2021
8802    2020
8803    2020
8804    2021
Name: work_year, Length: 8805, dtype: int64

In [5]:
salaries['work_year'] # hace lo mismo que el comando anterior pero evita problemas si las columnas tienen espacioes en el nombre

0       2023
1       2023
2       2023
3       2023
4       2023
        ... 
8800    2020
8801    2021
8802    2020
8803    2020
8804    2021
Name: work_year, Length: 8805, dtype: int64

In [6]:
salaries[['work_year']] # agregar un segundo set de corchetes hace que se muestre mas ordenado (en formato de dataframe) 

,work_year
0,2023
1,2023
2,2023
3,2023
4,2023
...,...
8800,2020
8801,2021
8802,2020
8803,2020


In [7]:
# se pueden cargar las columans que quieras, solo debes agregar el nommbre de todas las columnas a cargar separadas por coma 
salaries[['work_year', 'salary']] 

,work_year,salary
0,2023,212000
1,2023,190000
2,2023,35000
3,2023,35000
4,2023,245700
...,...,...
8800,2020,412000
8801,2021,151000
8802,2020,105000
8803,2020,100000


## Calculos basicos

In [8]:
salaries['salary_in_usd'].mean()  #Mean saca promedio de los datos (en este caso promedio de sueldo en usd)

np.float64(149488.26564452017)

In [9]:
salaries['salary_in_usd'].min() #Min() muestra valor mas bajo y su contraparte Max(), muestra el mas alto

15000

In [10]:
salaries['salary_in_usd'].max()

615201

In [11]:
salaries['salary_in_usd'].sum() #Suma los valores de la columna

np.int64(1316244179)

In [12]:
salaries.mean(numeric_only = True) #retorna el promedio de todos los datos de cada columna, filtrando solo datos numericos

work_year          2022.737422
salary           174728.701533
salary_in_usd    149488.265645
remote_ratio         38.693924
dtype: float64

### Desviacion estandar

In [13]:
salaries.columns #Revisamos columnas para ver con cual practicar

Index(['work_year', 'experience_level', 'employment_type', 'job_title',
       'salary', 'salary_currency', 'salary_in_usd', 'employee_residence',
       'remote_ratio', 'company_location', 'company_size'],
      dtype='object')

**Para este caso utilizare el salario** De esta forma compararemos los sueldos

In [20]:
salaries['salary_in_usd'].mean() # Calculamos el promedio y la desviación estándar para analizar la dispersión de los salarios respecto al valor medio


np.float64(149488.26564452017)

In [19]:
salaries['salary_in_usd'].std() #std da la desviacion estandar. 

64222.1050583918

***En este caso, gracias a los datos obtenidos, podemos comprobar que el sueldo promedio es de aproximadamente 149.488 dolares y que los salarios se desvian del salario promedio en aproximadamente 64.222 dolares***

### Count & Shape

La funcion **count()** cuenta la cantidad de valores no nulos en la columna o dataframe (util si necesitas saber cuantos datos validos hay) <br>
La funcion **shape** muestra el numero de filas y columnas del DataFrame, **incluyendo los nulos** (util para conocer el tamaño del dataset)

In [28]:
salaries.count() # se cuentan los valores no nulos del dataframe

work_year             8805
experience_level      8805
employment_type       8805
job_title             8805
salary                8805
salary_currency       8805
salary_in_usd         8805
employee_residence    8805
remote_ratio          8805
company_location      8805
company_size          8805
dtype: int64

In [26]:
salaries['salary_in_usd'].count() # Se cuentan solo los valores en la columna salary_in_usd

np.int64(8805)

In [27]:
salaries.shape # Se consultan las dimensiones del DataFrame (cuantas columnas y filas hay)

(8805, 11)

### Estadistica descriptiva con metodo Describe

La funcion **describe()** no da en el mismo output la desviacion estandar, conteo de datos, promedio, entre otras estadisticas y datos de utilidad para estudiar el dataset

In [31]:
salaries.describe() # Como podemos ver aqui nos da las estadisticas que hemos estado sacando para todas las columnas con valores numericos

,work_year,salary,salary_in_usd,remote_ratio
count,8805.000000,8.805000e+03,8805.000000,8805.000000
mean,2022.737422,1.747287e+05,149488.265645,38.693924
std,0.542484,4.560690e+05,64222.105058,48.068060
min,2020.000000,1.400000e+04,15000.000000,0.000000
25%,2023.000000,1.055000e+05,105000.000000,0.000000
50%,2023.000000,1.441000e+05,142200.000000,0.000000
75%,2023.000000,1.900000e+05,185900.000000,100.000000
max,2023.000000,3.040000e+07,615201.000000,100.000000


Como podemos ver en la tabla nos da en un solo output el conteo de datos no nulos, el promedio, el valor minimo, el maximo, entre otras estadisticas y datos en las columnas work year, salary, salary in usd y remote ratio. <br>
Para los numeros que tine un +03, +05, etc, se puede escribir el numero tal cual en una celda y te dara el numero completo, por ejemplo:

In [32]:
8.805000e+03

8805.0

**Los porcentajes mostrados en la tabla, corresponden a los cuartiles**, los cuales son los puntos de corte del dataset, dividiendolo en 4 partes <br>
¿Por que se llaman cuartiles si son solo 3 porcentajes? Por que los porcentajes son los puntos de corte, ya que para separar algo en 4 partes necesitas 3 cortes, el 0% y el 100% tambien son percentiles, pero no cuartiles, **los cortes son 25%, 50% (Mediana) y 75%**. <br> ***Los cuartiles dividen el conjunto de datos en cuatro grupos de igual tamaño, una vez que los valores han sido ordenados de menor a mayor.***

Adicionalmente, para sacar la moda, puedes usar la funcion **mode()**, la cual **devolvera una serie de numeros (varios numeros)**, ya que puede haber mas de una moda en un conjunto de datos, por ejemplo, si en los salarios hay 2.000 personas con sueldo de 80.000 usd y 2.000 personas con sueldo de 50.000, digamos que todos los demas sueldos son distintos para este ejemplo, pero el sueldo de 80 mil y el de 50 mil son los que mas se repiten y no solo eso, se repiten la misma cantidad de veces, por lo que en este caso, hay dos modas. <br>

***Nota: si deseas sacar solo la primera moda encontrada, por ejemplo, puedes poner un indice en la funcion mode(), por ejemplo "salaries['salary_in_usd'].mode()[0]"*** 

In [36]:
salaries['salary_in_usd'].mode()[0] # Con indice, como se ve aqui, solo me trajo la primera moda encontrada

np.int64(150000)

In [37]:
salaries['salary_in_usd'].mode() # sin indice, como se puede ver me devolvio las modas 0 y 150.000

0    150000
Name: salary_in_usd, dtype: int64

In [38]:
salaries.describe()

,work_year,salary,salary_in_usd,remote_ratio
count,8805.000000,8.805000e+03,8805.000000,8805.000000
mean,2022.737422,1.747287e+05,149488.265645,38.693924
std,0.542484,4.560690e+05,64222.105058,48.068060
min,2020.000000,1.400000e+04,15000.000000,0.000000
25%,2023.000000,1.055000e+05,105000.000000,0.000000
50%,2023.000000,1.441000e+05,142200.000000,0.000000
75%,2023.000000,1.900000e+05,185900.000000,100.000000
max,2023.000000,3.040000e+07,615201.000000,100.000000


En este caso, podemos interpretar gracias a los valores que el **25%** de los registros, tiene un salario de 105.000 USD o menos, el **50%** 142.200 USD o menos y el **75%** tiene un salario de 185.900 USD o menos. <br>

**Por que menos?** <br> 
***Porque un cuartil no representa a un grupo, sino un punto de corte. Por ende, ese valor (sea del cuartil que sea) indica el valor maximo que alcanza un cierto porcentaje de los datos***, En el caso del primer cuartil (25%), el sueldo máximo alcanzado por el 25% inferior de los registros es de 105.000 USD

In [40]:
salaries.describe(include='object')

,experience_level,employment_type,job_title,salary_currency,employee_residence,company_location,company_size
count,8805,8805,8805,8805,8805,8805,8805
unique,4,4,124,22,86,74,3
top,SE,FT,Data Engineer,USD,US,US,M
freq,6336,8762,2062,8006,7527,7576,7881


A la funcion "describe()" se le pueden introducir distintos parametros, por ejemplo, que tipo de dato incluir. <br> 
En la celda anterior, se le introdujo el parametro "include" y se le especifico que nos incluyera los datos de tipo objeto. <br>
Como resultado, la funcion nos arrojó en "count" el conteo de los datos tipo objeto, no nulos de cada columna con ese tipo de dato, asi mismo ahora nos incluye la fila "Unique" que nos indica cuantos datos unicos hay en cada columna con valores del tipo de dato objeto. Por otro lado, tambien nos agrego la fila "top", la cual indica el valor mas frecuente de la columna, ya por ultimo, tenemos la fila "freq" la cual nos indica la frecuencia, digase, cuantas veces se repitio ese valor mas frecuente que aparece en top

Un ejemplo de interpretacion de los datos entregados en una columna por la funcion, puede ser en la columna Job_Title, en esa columna podemos ver que se encontraron 8.805 valores no nulos y 124 valores (nombres de titulos profesionales) unicos, asi mismo, podemos concluir, que el titulo mas repetido entre los valores, es el titulo de Data Engineer, el cual se repite 2.062 veces en la columna job_title

Otro parametro que se le puede dar a la funcion describe, es el parametro "percentiles", este nos permite consultar los percentiles que nosotros deseemos. por ejemplo, digamos que en lugar de los percentiles 25% 50% 75% quiero los percentiles 20%, 60% y 90%, para esto, el codigo seria con nuestro ejemplo de salarios "salaries.describe(percentiles= [0.2, 0.6, 0.9])" <br><br>

**Nota: la funcion describe, independiente de los percentiles que le demos, siempre nos va a dar de igual forma el percentil del 50%, esto ya que el 50% es la mediana de los datos**

In [43]:
salaries.describe(percentiles=[0.2, 0.6, 0.9])

,work_year,salary,salary_in_usd,remote_ratio
count,8805.000000,8.805000e+03,8805.000000,8805.000000
mean,2022.737422,1.747287e+05,149488.265645,38.693924
std,0.542484,4.560690e+05,64222.105058,48.068060
min,2020.000000,1.400000e+04,15000.000000,0.000000
20%,2022.000000,9.500000e+04,95000.000000,0.000000
50%,2023.000000,1.441000e+05,142200.000000,0.000000
60%,2023.000000,1.600000e+05,160000.000000,0.000000
90%,2023.000000,2.397480e+05,234000.000000,100.000000
max,2023.000000,3.040000e+07,615201.000000,100.000000


### Value Counts

La funcion **value_counts()** permite contar cuantos registros hay por cada valor en la columna, por ejemplo, en la siguiente celda, utilice la funcion antes mencionada para contar cuantos registros hay en la columna de job_title por cada titulo, esto con el fin de contar cuantos empleados hay por cada titulo profesional.

In [48]:
salaries['job_title'].value_counts()

job_title
Data Engineer                     2062
Data Scientist                    1852
Data Analyst                      1322
Machine Learning Engineer          908
Applied Scientist                  258
                                  ... 
Deep Learning Researcher             1
Consultant Data Engineer             1
BI Data Engineer                     1
Managing Director Data Science       1
AWS Data Architect                   1
Name: count, Length: 124, dtype: int64

En este caso, podemos concluir que tenemos 2062 empleados con el titulo de Data Engineer y 908 machine learning engineers, lo que nos puede ayudar a concluir que el titulo profesional mas comun entre nuestros empleados es el de Data Engineer

La funcion **value_count()**, al igual que otras funciones de pandas, puede recibir varios parametros, los cuales son: <br>
**normalize**: permite ver proporciones y es de tipo bool <br>
**sort**: permite ordenar por frecuencia y es de tipo bool <br>
**ascending**: permite ordenar los datos en orden ascendente y es de tipo bool <br>
**bins**: permite agrupar valores numericos y puede ser de tipo int o list <br>
**dropna**: permite decidir si incluimos o no los valores NaN (vacio) y es de tipo bool

In [50]:
# A continuacion un ejemplo del parametro ascending cuando lo ponemos en true
salaries['job_title'].value_counts(ascending=True) 

job_title
Sales Data Analyst                     1
Data DevOps Engineer                   1
BI Data Engineer                       1
Deep Learning Researcher               1
Business Intelligence Consultant       1
                                    ... 
Applied Scientist                    258
Machine Learning Engineer            908
Data Analyst                        1322
Data Scientist                      1852
Data Engineer                       2062
Name: count, Length: 124, dtype: int64

In [54]:
# Otra cosa que se puede hacer con value_counts es darle varias columnas, en cuyo caso pandas calcula la frecuencia de cada combinación única 
# de valores entre dichas columnas, devolviendo una Serie con índice múltiple.
salaries[['job_title', 'salary_in_usd']].value_counts() 

job_title                        salary_in_usd
Data Engineer                    160000           68
                                 130000           63
Applied Scientist                136000           54
Data Analyst                     100000           50
Data Scientist                   160000           48
                                                  ..
Data Engineer                    172500            1
                                 172600            1
                                 172800            1
                                 172900            1
Staff Machine Learning Engineer  185000            1
Name: count, Length: 3600, dtype: int64

***Con estos resultados, podemos concluir que por ejemplo, existen 68 registros donde el cargo del empleado es data engineer y el salario es exactamente 160.000 USD, lo que nos da a entender que la mayoria de empleados con el cargo de data engineer tiene un sueldo de 160.000 USD***